### ISSUE TITLE
test_imap flaky with randomized hashes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5750]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#5750|
|Type|defect|
|Created|2012-07-08 22:24:19Z|

One test depends on hash ordering and fails sporadically with ``python -R``:

```
[FAIL]
Traceback (most recent call last):
  File "/home/antoine/twisted/twisted/mail/test/test_imap.py", line 357, in test_headerFormatter
    self.assertEqual(imap4._formatHeaders(input), output)
  File "/home/antoine/twisted/twisted/trial/unittest.py", line 270, in assertEqual
    % (msg, pformat(first), pformat(second)))
twisted.trial.unittest.FailTest: not equal:
a = 'Header1: Value1\r\nHeader2: Value2\r\n'
b = 'Header2: Value2\r\nHeader1: Value1\r\n'
```

Attachments:

* [randomhash_test_imap.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b7f/b7f58bdc03119612000fff9c609acdc4de598c2d/bec122f771bd8f96aabed7f31f2db486fc87e1f8.patch) (688 bytes) - added by antoine on 2012-07-08 22:24:41Z - 
* [randomhash_test_imap2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b7f/b7f58bdc03119612000fff9c609acdc4de598c2d/55c6038f0b4d431d5f5d04f63c7e562785c58943.patch) (2353 bytes) - added by antoine on 2012-07-08 22:44:57Z - 
* [randomhash_test_imap3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b7f/b7f58bdc03119612000fff9c609acdc4de598c2d/b7065b7f97c1d17a70c7fcd5b31092f46c8591a2.patch) (688 bytes) - added by antoine on 2012-07-24 21:45:08Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5750 5750
type__defect defect
reporter__antoine antoine
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1341786259000000 1341786259000000
changetime__1344343125000000 1344343125000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_1"></a>|@pitrou commented|
|-|-|

Patch attached.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_2"></a>|@pitrou commented|
|-|-|

Hmm, sorry, other tests are also failing:

```
[FAIL]
Traceback (most recent call last):
  File "/home/antoine/twisted/twisted/mail/test/test_imap.py", line 3412, in <lambda>
    d.addCallback(lambda x : self.assertEqual(self.result, self.expected))
  File "/home/antoine/twisted/twisted/trial/unittest.py", line 270, in assertEqual
    % (msg, pformat(first), pformat(second)))
twisted.trial.unittest.FailTest: not equal:
a = {0: {'BODYSTRUCTURE': ['text',
                       'plain',
                       [['key', 'value'], ['name', 'thing']],
                       'this-is-the-content-id',
                       'describing-the-content-goes-here!',
                       '8BIT',
                       '20',
                       '4',
                       None,
                       None,
                       None]}}
b = {0: {'BODYSTRUCTURE': ['text',
                       'plain',
                       [['name', 'thing'], ['key', 'value']],
                       'this-is-the-content-id',
                       'describing-the-content-goes-here!',
                       '8BIT',
                       '20',
                       '4',
                       None,
                       None,
                       None]}}


twisted.mail.test.test_imap.NewFetchTestCase.testFetchBodyStructure
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/home/antoine/twisted/twisted/mail/test/test_imap.py", line 3412, in <lambda>
    d.addCallback(lambda x : self.assertEqual(self.result, self.expected))
  File "/home/antoine/twisted/twisted/trial/unittest.py", line 270, in assertEqual
    % (msg, pformat(first), pformat(second)))
twisted.trial.unittest.FailTest: not equal:
a = {0: {'BODYSTRUCTURE': ['text',
                       'plain',
                       [['key', 'value'], ['name', 'thing']],
                       'this-is-the-content-id',
                       'describing-the-content-goes-here!',
                       '8BIT',
                       '20',
                       '4',
                       None,
                       None,
                       None],
     'UID': '919293'}}
b = {0: {'BODYSTRUCTURE': ['text',
                       'plain',
                       [['name', 'thing'], ['key', 'value']],
                       'this-is-the-content-id',
                       'describing-the-content-goes-here!',
                       '8BIT',
                       '20',
                       '4',
                       None,
                       None,
                       None],
     'UID': '919293'}}


twisted.mail.test.test_imap.NewFetchTestCase.testFetchBodyStructureUID
```

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_3"></a>|@pitrou commented|
|-|-|

Ok, the latest patch (randomhash_test_imap2.patch) fixes the whole of test_imap. It also removes an intermediate dict in getBodyStructure().

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @pitrou|
|-|-|

Thanks.  There is no test coverage of the last hunk in this diff, the part that changes `result.append(attrs.items())` to `result.append(attrs)`.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_6"></a>|@pitrou commented|
|-|-|

Replying to [exarkun](#note_5):
> Thanks.  There is no test coverage of the last hunk in this diff, the part that changes `result.append(attrs.items())` to `result.append(attrs)`.

Ok, not only there is no coverage, but I found that getBodyStructure's behaviour with multipart messages is probably buggy (as hinted in the function comments - "![XXX - This does not properly handle multipart messages](/uploads/migrated/XXX - This does not properly handle multipart messages)" :-)).
Opened ticket 5763 to track this (with a patch).

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_7"></a>|@pitrou commented|
|-|-|

New patch without the bodystructure fixes, which aren't necessary anymore.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)|@pitrou removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  This seems sensible.  The modified test method would benefit from a docstring, of course. :)  I'll add that and a news fragment and apply, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [35084]) Remove dependence on dictionary iteration order from an IMAP4 unit test.

Author: antoine
Reviewer: exarkun
Fixes: [#5750](https://github.com/twisted/twisted/issues/5750)
